src/benchmark/tpch_benchmark.cpp

```diff
@@ -0,0 +1,355 @@
+   public:
+    int overflow(int c) { return c; }
+  };
+
+  static NullBuffer null_buffer;
+  static std::ostream null_stream(&null_buffer);
+
+  return null_stream;
+}
+
+struct QueryBenchmarkResult {
+  size_t num_iterations = 0;
+  Duration duration = Duration{};
+};
+
+using BenchmarkResults = std::unordered_map<size_t, QueryBenchmarkResult>;
+
+/**
+ * Loosely copying the functionality of benchmark::State
+ * keep_running() returns false once enough iterations or time has passed.
+ *
+ */
+struct BenchmarkState {
+  enum class State { NotStarted, Running, Over };
+
+  BenchmarkState(const size_t max_num_iterations, const Duration max_duration)
+      : max_num_iterations(max_num_iterations), max_duration(max_duration) {}
+
+  bool keep_running() {
+    switch (state) {
+      case State::NotStarted:
+        begin = std::chrono::high_resolution_clock::now();
+        state = State::Running;
+        break;
+      case State::Over:
+        return false;
+      default: {}
+    }
+
+    if (num_iterations >= max_num_iterations) {
+      end = std::chrono::high_resolution_clock::now();
+      state = State::Over;
+      return false;
+    }
+
+    end = std::chrono::high_resolution_clock::now();
+    const auto duration = end - begin;
+    if (duration >= max_duration) {
+      state = State::Over;
+      return false;
+    }
+
+    num_iterations++;
+
+    return true;
+  }
+
+  State state{State::NotStarted};
+  TimePoint begin = TimePoint{};
+  TimePoint end = TimePoint{};
+
+  size_t num_iterations = 0;
+  size_t max_num_iterations;
+  Duration max_duration;
+};
+
+// Stores config and state of a benchmark mode instantiation
+class TpchBenchmark final {
+ public:
+  TpchBenchmark(const BenchmarkMode benchmark_mode, const std::optional<std::string>& output_file_path,
+                std::vector<size_t> query_ids, const opossum::ChunkOffset chunk_size, const float scale_factor,
+                const size_t max_num_query_runs, const Duration max_duration)
+      : _benchmark_mode(benchmark_mode),
+        _output_file_path(output_file_path),
+        _query_ids(std::move(query_ids)),
+        _chunk_size(chunk_size),
+        _scale_factor(scale_factor),
+        _max_num_query_runs(max_num_query_runs),
+        _max_duration(max_duration),
+        _query_results_by_query_id() {}
+
+  void run() {
+    /**
+     * Populate the StorageManager with the TPC-H tables
+     */
+    out() << "- Generating TPCH Tables with scale_factor=" << _scale_factor << "..." << std::endl;
+    opossum::TpchDbGenerator(_scale_factor, _chunk_size).generate_and_store();
+    out() << "- Done." << std::endl;
+
+    /**
+     * Run the TPCH queries in the selected mode
+     */
+    if (_benchmark_mode == BenchmarkMode::IndividualQueries) {
+      _benchmark_individual_queries();
+    } else if (_benchmark_mode == BenchmarkMode::PermutedQuerySets) {
+      _benchmark_permuted_query_sets();
+    } else {
+      Fail("Invalid mode");
```

Comment:
Why not `switch`? That would handle the "invalid mode" case automatically.

Reply:
yup, since we have `-Wswitch`, switch actually seems better

Reply:
https://clang.llvm.org/docs/DiagnosticsReference.html#wswitch